### DISCUSSION TITLE
Lazy loading tiles

### DISCUSSION THREAD

**Discussion by clairemibo** [valhalla/valhalla#3323] [Q&A] [answered]
Hi,

Thanks for taking time reviewing my question!

We have built Valhalla Service to serve navigation requests. Currently, the way it loads tiles is to download the tiles tar file (from some remote storage) on service start. The planet tar is really huge in size (70G), so the time for downloading is about 10 mins or so. In the case we need to do a hotfix for the service, we'd need to restart the service and thus there'll be 10 mins for not serving requests. 

We'd like to change it to a lazy loading style - something like this - we don't download the tiles on service start, when a request comes in, figure out which tiles are needed, if they're not in memory or on disk, fetch them from some remote storage, save them on disk. 

Is this idea feasible? Is there any suggestion on how we can make this happen?

Thanks!

**Comment by kevinkreiser** [comment, marked as answer]
@clairemibo this is already supported. you cannot use the tiles tar to achieve this but you must instead use the `tile_dir` configuration. you can indeed configure this to point to an empty directory. then you also need to set up an http endpoint from which valhalla can fetch tiles when it needs them. we cannot precompute what tiles we will need for a route, that isnt how routing works, though you could write a heuristic that might be good at guessing. still even if you had this you would have to synchronize your prefetching with the actual route computation which sounds doable but pretty annoying. anyway, instead of all that we pull them on the fly as the algorithm decides it needs them. they are then cached into the tile_dir and wont be fetched over network again. please look at `valhalla_build_config` to see all the options you have in there for tilecache and data config

also note that there are drawbacks to this approach:

1. cold cache is very slow because now you have network hops in your tight loop
2. if you want to "hotswap" your data you need to clear your cache anyway because you cant mix old and new tiles

@nilsnolde and i are working on the ability to reload a tar without restarting. at the moment it wouldnt be enough to achieve what you are asking for but it is the first and arguably most important step. you can see more at #3281

**Reply by coogle** [reply]
@kevinkreiser @nilsnolde 

I'm trying to find documentation for doing something like this (hosting an HTTP endpoint with tile data, and requesting as needed instead of downloading the tile data and storing locally) but I can't seem to find anything. Can you point me to some docs on how this would work?

**Reply by kevinkreiser** [reply]
its not really documented except in the config file: https://github.com/valhalla/valhalla/blob/40e898e5025a5be479fc622b10f730673e6ddcc4/scripts/valhalla_build_config#L294

so when you make tiles with valhalla_build_tiles the output is a directory with gph files in it. if you were to serve that directory as the root with a webserver and point that config value at that webserver it should just work.

**Reply by coogle** [reply]
Ah great, that's what I needed to know -- can I literally just toss the tiles directory on a static webserver and point the URL. Thanks!

**Reply by kevinkreiser** [reply]
yep

**Reply by coogle** [reply]
Is there any sort of auto-cleanup that happens with this, or in theory would you eventually end up with a 70GB `tile_dir` directory over time as you continued to download tiles from the webserver?

In my use case I am willing to trade speed for a smaller footprint, so I'm trying to decide if I should perhaps just nuke the `tile_dir` after every request is fulfilled (using it in one-shot mode here).

**Reply by kevinkreiser** [reply]
at the moment we dont purge tiles on the filesystem (just in memory) however it wouldnt be too hard to tie this into the existing LRU implementation

**Comment by clairemibo** [comment]
Thanks Kevin!
